soc/arm/rpi_pico/rp2/soc.c

```diff
@@ -30,6 +33,17 @@ extern void z_arm_nmi_init(void);
 
 LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
 
+/* Overrides the weak ARM implementation:
+   Set general purpose retention register and reboot */
+void sys_arch_reboot(int type)
+{
+	if (type != 0) {
```

Comment:
Use `SYS_REBOOT_WARM`

Reply:
Does that really match? I know that nRF52840 uses a different special code shoved into a register that various bootloaders will honor to enter the bootloader, which seems a bit different than a normal "warm reboot".

Reply:
What are you using this for? As far as I know `type` is either warm or cold

Reply:
The nrf52840 version of this sets a specific retained register to whatever value is passed in:

```
void sys_arch_reboot(int type)
{
	nrf_power_gpregret_set(NRF_POWER, (uint8_t)type);
	NVIC_SystemReset();
}
```

The adafruit bootloader inspects that register for a specific value, and if detected, enters the bootloader instead of launching the already flashed application.

Ideally, we should be able to replicate that *somehow* on RP2040. I don't see *any* consistent conventions for other SoCs, nor any explicit note about the reboot type being extensible. From `include/zephyr/sys/reboot.h`:

```
#define SYS_REBOOT_WARM 0
#define SYS_REBOOT_COLD 1

/**
 * @brief Reboot the system
 *
 * Reboot the system in the manner specified by @a type.  Not all architectures
 * or platforms support the various reboot types (SYS_REBOOT_COLD,
 * SYS_REBOOT_WARM).
 *
 * When successful, this routine does not return.
 */
extern FUNC_NORETURN void sys_reboot(int type);
```

@gmarull Any suggestions for us?